benchmark/benchmark-pandoc.hs

```diff
@@ -21,39 +20,49 @@ Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 import Prelude
 import Text.Pandoc
 import Text.Pandoc.MIME
-import Text.Pandoc.Error (PandocError(..))
-import Control.Monad.Except (throwError)
+import Control.Monad.Except (throwError, liftIO)
 import qualified Text.Pandoc.UTF8 as UTF8
 import qualified Data.ByteString as B
 import qualified Data.Text as T
 import Criterion.Main
 import Criterion.Types (Config(..))
 import Data.List (intersect)
-import Data.Maybe (mapMaybe)
+import Data.Maybe (mapMaybe, catMaybes)
 import System.Environment (getArgs)
 import qualified Data.ByteString.Lazy as BL
 
+data Input = InputText {unInputText :: T.Text}
+           | InputBS {unInputBS :: BL.ByteString}
+
 readerBench :: Pandoc
             -> T.Text
-            -> Maybe Benchmark
-readerBench doc name =
-  case res of
-       Right (readerFun, inp) ->
-          Just $ bench (T.unpack $ name <> " reader")
-               $ nf (\i -> either (error . show) id $ runPure (readerFun i))
-                 inp
-       Left _ -> Nothing
-  where res = runPure $ do
-          (rdr, rexts) <- getReader name
+            -> IO (Maybe Benchmark)
+readerBench doc name = do
+  let (rdr, rexts) = either (error . show) id . runPure $ getReader name
+  res <- runIO $ do
           (wtr, wexts) <- getWriter name
           case (rdr, wtr) of
             (TextReader r, TextWriter w) -> do
-                     setResourcePath ["../test"]
+                     setResourcePath ["./test"]
```

Comment:
Was this just a mistake in the original, or did you change something that makes this necessary?

Reply:
This was apparently a mistake in the original, since various `cabal bench` run benchmarks with the root directory as `cwd`. For pure writers it doesn't really matter though, since those don't access the FS anyway.